testnet/stacks-node/src/nakamoto_node/signer_coordinator.rs

```diff
@@ -0,0 +1,436 @@
+    weight_threshold: u32,
+    /// Tracks signatures for blocks
+    ///   - key: Sha512Trunc256Sum (signer signature hash)
+    ///   - value: BlockStatus
+    blocks: Arc<(Mutex<HashMap<Sha512Trunc256Sum, BlockStatus>>, Condvar)>,
+    /// Tracks the timestamps from signers to decide when they should be
+    /// willing to accept time-based tenure extensions
+    ///  - key: StacksPublicKey
+    ///  - value: TimestampInfo
+    signer_idle_timestamps: Arc<Mutex<HashMap<StacksPublicKey, TimestampInfo>>>,
+    /// Keep running flag for the signer DB listener thread
+    keep_running: Arc<AtomicBool>,
+    /// Handle for the signer DB listener thread
+    listener_thread: Option<JoinHandle<()>>,
+}
+
+impl SignerCoordinator {
+    /// Create a new `SignerCoordinator` instance.
+    /// This will spawn a new thread to listen for messages from the signer DB.
+    pub fn new(
+        stackerdb_channel: Arc<Mutex<StackerDBChannel>>,
+        node_keep_running: Arc<AtomicBool>,
+        reward_set: &RewardSet,
+        burn_tip: &BlockSnapshot,
+        burnchain: &Burnchain,
+        message_key: StacksPrivateKey,
+        config: &Config,
+    ) -> Result<Self, ChainstateError> {
+        info!("SignerCoordinator: starting up");
+        let keep_running = Arc::new(AtomicBool::new(true));
+
+        // Create the stacker DB listener
+        let mut listener = StackerDBListener::new(
+            stackerdb_channel,
+            node_keep_running.clone(),
+            keep_running.clone(),
+            reward_set,
+            burn_tip,
+            burnchain,
+        )?;
+        let is_mainnet = config.is_mainnet();
+        let rpc_socket = config
+            .node
+            .get_rpc_loopback()
+            .ok_or_else(|| ChainstateError::MinerAborted)?;
+        let miners_contract_id = boot_code_id(MINERS_NAME, is_mainnet);
+        let miners_session = StackerDBSession::new(&rpc_socket.to_string(), miners_contract_id);
+
+        let mut sc = Self {
+            message_key,
+            is_mainnet,
+            miners_session,
+            total_weight: listener.total_weight,
+            weight_threshold: listener.weight_threshold,
+            blocks: listener.blocks.clone(),
+            signer_idle_timestamps: listener.signer_idle_timestamps.clone(),
+            keep_running,
+            listener_thread: None,
+        };
+
+        // Spawn the signer DB listener thread
+        let listener_thread = std::thread::Builder::new()
+            .name("stackerdb_listener".to_string())
+            .spawn(move || {
+                if let Err(e) = listener.run() {
+                    error!("StackerDBListener: exited with error: {e:?}");
+                }
+            })
+            .map_err(|e| {
+                error!("Failed to spawn stackerdb_listener thread: {e:?}");
+                ChainstateError::MinerAborted
+            })?;
+
+        sc.listener_thread = Some(listener_thread);
```

Comment:
One of these days, we need to centralize all the singleton thread-creation that happens in the node.

Reply:
Resolving this conversation for now.